feat(secondary-nav): add `<rh-secondary-nav>`

zeroedin opened the pull request:
## `<rh-secondary-nav>`

## What I did

1. Implement `<rh-secondary-nav>` and child components
2. Add `variant=dark` attribute

Fixes: #236

## Testing instructions 

1. `npm run test` will perform available unit tests for this component

## Notes for Reviewers

1. Running `npm run start` will display the secondary nav in an iframe like view.  In order to see the correct breakpoints, you will need to use the demo's fullscreen view.  **Caveat:** Full-screen mode breaks the `position: sticky` so menu will no longer stick to the top on scroll.  
2. Running `npm run proxy rh-secondary-nav` in a second terminal window will allow you to view the secondary nav in context with redhat.com, after you run the command it will tell you what port it is running on ie: http://localhost:8001 this will allow you to view with `position: sticky` and a generalized idea of implementation of how it could overlay primary nav + banners/universal-nav above.  Doing so will also provide a better showing of how secondary nav overlays over real content when a dropdown is clicked and scroll behavior on mobile and short screen sizes.

**TODO:**
- [x] add overlay on dropdown menu open, missed that in design spec
- [x] clean up dropdown state management between mobile and desktop views 
- [x] ensure when focus is outside element its state returns to closed.
- [x] mobile testing
- [x] write spec tests
- [x] write inline documentation / JSDoc
- [x] accessibility review
- [x] design review
- [x] code review

bennypowers commented:
hey @zeroedin you know `<dialog>` doesn't trap focus if it's not opened with `showModal()`?

try it on https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

zeroedin commented:
@coreyvickery Committed fixes for your requested changes/fixes.  The alignment issue is still unresolved and open for further discussion.

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+ * @slot links            - Only for type="fixed-width", expects a `<ul>`
+ * @slot cta              - Menu level CTA, expects a `<pfe-cta>`
+ *
+ * @csspart full-width    - {HTMLElement} container - The wrapper for full-width menus
```
```suggestion
 * @csspart full-width    - container - The wrapper for full-width menus
```
and others. `@csspart` doesn't take a type, the analyzer will just spit out a description string with `{HTMLElement}` in it, which isn't great.

zeroedin replied:
Fixed in 2f88b975241399448e917d6be39fb408e4026d29

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+  static readonly styles = [styles];
+
+  // TODO: change to variant instead of type which is used by html element properties such as input.
+  @property({ reflect: true }) type: 'fixed-width' | 'full-width' = 'full-width';
```
in this case what about `layout` or `size`?

zeroedin replied:
`layout` is most likely the most appropriate IMHO.  

The `TODO:` comment was a holdover from prior to posting the discussion https://github.com/RedHat-UX/red-hat-design-system/discussions/293 on this very question.  I am open and would like to hear more opinions on this specific instance from the designers and will reference this comment there before making a change.

bennypowers replied:
if you ask me, go with `layout` for now, and if we need to change it later per design's input, we'll do that in a subsequent PR

zeroedin replied:
Fixed in e441b00b80889be1fa6635a336fbf011e64c5334

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+  @property({ reflect: true }) type: 'fixed-width' | 'full-width' = 'full-width';
+
+  @queryAssignedNodes('cta', true)
+  private _ctaNodes: NodeListOf<HTMLElement> | undefined;
```
```suggestion
  private _ctaNodes?: NodeListOf<HTMLElement>;
```
equivalent, a drop terser, and to my eye nicer. season to taste

zeroedin replied:
Fixed in a90891b0d89c88bb1e42d7c922099a606c44cf7b

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+   * when viewport changes at breakpoint or first load of the component.
+   */
+  @observed
+  @property({ reflect: false }) _isMobile = false;
```
```suggestion
  @state() private _isMobile = false;
```

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+  /**
+   * Updates isMobile property with matchMedia when viewport changes
+   */
+  protected matchMedia = new MatchMediaController(this, `(min-width: ${tabletLandscapeBreakpoint})`, {
```
can we use RHDSScreenSizeController here instead, so that we don't have to import tokens here? you also wouldn't have to set a state property either, just 
```ts
const { mobile } = this.#screenSize;
return html`<div class=${classMap({ mobile })}></div>`;
```

zeroedin replied:
That's muy caliente 🌶️ !  Yes definitely!

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+  render() {
+    const classes = { 'is-mobile': this._isMobile };
+
+    let menuContents;
```
can we inline this?
```js
return html`
  <div id="container" class="${classMap(classes)}">${this.type === 'full-width' ? html`
    <div id="full-width" part="full-width">
      <slot name="base">
        <div id="sections" part="sections" hidden>
          <slot name="section" @slotchange=${this.#onSectionsSlotChange}></slot>
        </div>
        <div id="cta" part="cta" hidden>
          <slot name="cta" @slotchange=${this.#onCtaSlotChange}></slot>
        </div>
      </slot>
    </div>` : html`
    <div id="fixed-width" part="fixed-width">
      <slot name="base">
        <slot name="links"></slot>
      </slot>
    </div>`}
  </div>`;
  ```

zeroedin replied:
Fixed in dae47d21f1dbb493c0ea02097e84efb399fe5e64

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+   */
+  toggleVisibility(visible: boolean) {
+    if (visible) {
+      this._container?.classList.add('visible');
```
see previous, wrt imperative updates of shadow dom

zeroedin replied:
Fixed in 1c4690143e173a313836f8b8944be2cb8fd11efc

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+  #fullWidthMenu() {
+    return html`
+      <div id="full-width" part="full-width">
+        <slot name="base">
```
see prev re `base`

zeroedin replied:
Fixed in dae47d21f1dbb493c0ea02097e84efb399fe5e64

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+    if (!this._ctaNodes) {
+      return;
+    }
+    const hasCta = this._ctaNodes.length > 0;
```
```suggestion
    this._cta?.toggleAttribute('hidden', this._ctaNodes.length > 0);
```
https://developer.mozilla.org/en-US/docs/Web/API/Element/toggleAttribute#parameters

bennypowers replied:
you could also avoid the `if(p) return` by nullish coalescing to 0 with optional chaining on the `toggleAttribute` call

bennypowers replied:
but in any case, we should prefer to template this based on component state, rather than doing imperative updates

zeroedin replied:
Cleaned up a bit in f5c0101068cec480c8513addb9eb25fb85abe708.  However, I would like to talk about the "preferred" method more.  I might be missing some points on how we can use the component state as it relates to checking if the slot has slotted children.  If there is another method I would love to learn it.

zeroedin replied:
Fixed in 4e6a18a554cbf20c1268d594fdadca59cf0d9b94

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-menu.ts:
```diff
@@ -0,0 +1,163 @@
+      return;
+    }
+    const hasSections = this._sectionsNodes.length > 0;
+    hasSections ? this._sections?.removeAttribute('hidden') : this._sections?.setAttribute('hidden', '');
```
see prev

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-overlay.ts:
```diff
@@ -0,0 +1,62 @@
+  @state()
+  public open = false;
+
+  connectedCallback() {
```
remove

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-overlay.ts:
```diff
@@ -0,0 +1,62 @@
+    super.connectedCallback();
+  }
+
+  render() {
```
remove

zeroedin replied:
Fixed in 61bea434ccfe42654370827ed28825074a8c6f9d

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav-overlay.ts:
```diff
@@ -0,0 +1,62 @@
+   * @param parent {RhSecondaryNav} parent nav of overlay
+   */
+  @bound
+  public toggleNavOverlay(toggle: HTMLElement, state: boolean, parent: RhSecondaryNav) {
```
rather than making this an imperative method call, why not just set `.open = true`, then handle the side effects here (or even in css)
```ts
@property({ type: Boolean, reflect: true }) open = false;
```
```css
:host(:not([open])) {
  display: none;
}
```

zeroedin replied:
Fixed in a4fa730f5058f56e4e829a9938b5264bf25a2b8d

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.css:
```diff
@@ -0,0 +1,147 @@
+  flex-direction: column;
+  padding: 2em 1em 1em;
+  margin: 0 !important;
+  max-height: calc(100vh - var(--rh-secondary-nav-min-height ,4.625em));
```
```suggestion
  max-height: calc(100vh - var(--rh-secondary-nav-min-height, 4.625em));
```

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.css:
```diff
@@ -0,0 +1,147 @@
+  display: flex !important;
+  list-style: none;
+  flex-direction: column;
+  padding: 2em 1em 1em;
```
caveat, tokens

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.css:
```diff
@@ -0,0 +1,147 @@
+  display: flex;
+  height: 100%;
+  align-items: center;
+  font-family: var(--pfe-theme--font-family);
```
warning, tokens

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.css:
```diff
@@ -0,0 +1,147 @@
+}
+
+
+button:after {
```
```suggestion
button::after {
```

zeroedin replied:
Fixed in 190153fba40663184149e3b5a41720e7b49f1dd6

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+   * when viewport changes at breakpoint or first load of the component.
+   */
+  @observed
+  @property({ reflect: false }) _isMobile = false;
```
see previous re RHDSScreenSizeController

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.addEventListener('change', this._dropdownChangeHandler as EventListener);
```
in this case i recommend sticking with lib's typing and do an `instanceof` check inside the handler. if a user slots a `select` or `input` element and it changes...

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+  firstUpdated() {
+    // after update the overlay should be available to attach an event listener to
+    this._overlay?.addEventListener('click', this._overlayClickHandler as EventListener);
+    this._isMobile = (window.outerWidth < parseInt(tabletLandscapeBreakpoint.toString().split('px')[0]));
```
if this is needed, i'd rather see it calculated in `updated`, or in a (new) callback in RHDSScreenSizeController

bennypowers replied:
My issue here isn't that the initial implementation of `RHDSScreenSizeController` is sacrosanct or anything - by all means improve it! My only issue here is doing this logic inside the component. If you need to make changes to `RHDSScreenSizeController` to get this feature in secondary-nav, go for it, I just don't want to see breakpoints imported or hard coded into components.

The reason why this is important is so that we maintain some level of uniformity and control over the design system. This is also the reason we created the tokens repo. We could very easily loosen this discipline, but that would lead to a situation where each individual component[ author] makes their own design assumptions, and keeps them private and local to line N in whichever component file. Even in the short term, this degrades maintainability.

So to answer your question in the google doc, yes, we should eventually put those components in the tokens repo and import them into `RHDSScreenSizeController.ts`. If, for the time being, you need to hard code values there or import them from `lib/tokens.ts`, no problemo, it's still less to worry about than having those token references scattered about in component lifecycle methods.

> But secondary-nav uses values not represented in tokens repo or tokens.ts

Add those values there. Components should not own design tokens.

> But I need to get this out for secondary nav now

Add the values in `RHDSScreenSizeController.ts`, in this PR. Make sure it doesn't break table. This shouldn't be a blocker, from what you've described, the problem is the internal implementation of `RHDSScreenSizeController`, which is very much open to change and improvement.

My _only_ requirements here are that breakpoints be handled exclusively in the controller, and that the component APIs be declarative (i.e., that components aren't doing any comparisons against token values). This is so that tokens are
1. available to any component while being owned by none
2. defined (if even initially aspirationally/locally) in the tokens
3. themeable

zeroedin replied:
Fixed in 7996deb4adcfb81f26ebf40d05427e5bb55abb5c and c03d4820dbd8a6b95ee2520b3a7f5c4968b6e766

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+   * @return {void}
+   */
+  @bound
+  private _dropdownChangeHandler(event: SecondaryNavDropdownChangeEvent): void {
```
🤔 thinking this over

if you want to keep the parent in charge of setting state, would it make sense to call the event `expand-request`?

zeroedin replied:
Assuming this is also just a naming issue, I have no problem switching that to expand-request.  It does give the event a more specific meaning.  

As far as wanting the parent in charge,  I assumed the parent would be the best place to handle the state of its children's dropdowns.  More specifically that any new dropdown open event closes all other open dropdowns so I just followed the parent -> child model thought process. 

You had mentioned at some point the idea of possibly abstracting the dropdowns (flyout menus)  into a completely separate component.  I haven’t thought about how that change would impact how I'm currently handling the open-close here.  

I can potentially see dropdowns listening for other dropdowns open request events, and if one is open all others close in response.  For reusability, If they aren’t used in a “grouping” like in the navigation menu itself, maybe you might not want that functionality. You might want to allow more than 1 to be open at a time, such as in an accordion, maybe?  Thought for later.  

That said, would something like that approach be preferable?  The dropdowns listen for `expand-request` events, and if it wasn’t the dropdown that called the event it would change its own state to closed?

zeroedin replied:
Fixed in a14bf72a07b3519ddbe0c9e65ea0840f047dc53f

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+   * @param newVal {boolean | undefined}
+   * @returns {void}
+   */
+  private __isMobileChanged(oldVal?: boolean | undefined, newVal?: boolean | undefined): void {
```
perhaps best handled with a (new) effects callback on screensize controller

zeroedin replied:
Fixed in c03d4820dbd8a6b95ee2520b3a7f5c4968b6e766

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -0,0 +1,389 @@
+    if (this._mobileMenuButton?.getAttribute('aria-expanded') === 'true') {
+      return;
+    }
+    this._mobileMenuButton?.setAttribute('aria-expanded', 'true');
```
prefer to template this, using private state if necessary

bennypowers commented on elements/rh-secondary-nav/test/rh-secondary-nav.spec.ts:
```diff
@@ -0,0 +1,185 @@
+      });
+
+      it('lightdom passes the a11y audit', async function() {
+        await expect(element).to.be.accessible();
```
i don't believe you need to run this on both the element and it's `shadowDom`. Just `expect(element).to.be.accessible()` is sufficient

zeroedin replied:
Fixed in bd4ee5b939bc72ff81264f416e1a76b5ac421e2d

bennypowers commented on elements/rh-secondary-nav/test/rh-secondary-nav.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    expect(element.hasAttribute('role')).to.be.false;
+  });
+
+  it('should have an overlay set to hidden after upgrade', async () => {
```
```suggestion
  it('should have an overlay set to hidden after upgrade', async function() {
```

prefer [function keyword in mocha tests](https://mochajs.org/#arrow-functions)

zeroedin replied:
Fixed in: de6c9fedc09782910db0fe477bb6879c55876c26

bennypowers commented on elements/rh-secondary-nav/demo/rh-secondary-nav.html:
```diff
@@ -0,0 +1,414 @@
+  </a>
+  <ul slot="nav" aria-labelledby="logo-id"> <!-- is this ul labelledby the text or above what if its a svg? -->
+    <li>
+      <rh-secondary-nav-dropdown>
```
I've got this itch in the back of my neck that says "pssst... abstract this into `/elements`..."

but also YAGNI, we'll worry about that when the time comes

zeroedin replied:
Can you explain this a bit more?  Are you suggesting getting rid of the `<rh-secondary-nav-dropdown>` element and abstracting it into the shadow dom so to speak?  If so let's have a conversation on that.  That should be something i should be able to tackle.

bennypowers commented on elements/rh-secondary-nav/demo/rh-secondary-nav.html:
```diff
@@ -0,0 +1,414 @@
+              <li>
+                <a href="#">
+                  Integrations
+                  <pfe-icon icon="web-icon-new-window" color="info" size="sm"></pfe-icon>
```
beware eventual deprecation of pfe-icon or replacement with rh-icon

zeroedin replied:
`<pfe-icon>` in that case is light dom and could be replaced by anything.

bennypowers commented on elements/rh-secondary-nav/demo/rh-secondary-nav.html:
```diff
@@ -0,0 +1,414 @@
+      <rh-secondary-nav-dropdown>
+        <a href="#" slot="link">Grids</a>
+        <rh-secondary-nav-menu slot="menu">
+          <div slot="base"><h3>Grid</h3></div>
```
I dunnoo..... did people ask for this? I'd rather encourage them to use `::part()` and grid tokens, tbh

zeroedin replied:
Fixed in dae47d21f1dbb493c0ea02097e84efb399fe5e64

bennypowers reviewed (changes requested):
Code looks neat and organized, it's clear the care and thought that went into this and the results speak for themselves

My comments here are divided into three categories:
0. caveats about future (internal) api changes
1. nitpicks
2. proper change requests
you should be able to suss out which are which. ping here or DM with qs.

Nice work!

zeroedin commented on elements/rh-secondary-nav/demo/demo.css:
```diff
@@ -0,0 +1,23 @@
+:root {
+  --pfe-theme--font-family: "Red Hat Text", "Overpass", Helvetica, Arial, sans-serif;
```
I'm actively looking forward to cleaning up some of the CSS when they do!

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.css:
```diff
@@ -0,0 +1,56 @@
+  gap: 0.85em;
+}
+
+::slotted([slot="link"]):after {
```
old dog 🐶 bad habits 😆.  Good catch.

zeroedin replied:
Fixed in 190153fba40663184149e3b5a41720e7b49f1dd6

bennypowers commented:
@zeroedin please also add yourself to the contributors list in package.json

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+ *
+ * @fires { SecondaryNavDropdownChangeEvent } - Fires when a dropdown is clicked
+**/
+@customElement('rh-secondary-nav-dropdown') @pfelement()
```
Fixed in 0507600cb628ab536778cd1ab2bda775da1ff888

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+ * @slot link   - Link for dropdown, expects `<a>`
+ * @slot menu   - Menu for dropdown, expects `<rh-secondary-nav-menu>`
+ *
+ * @fires { SecondaryNavDropdownChangeEvent } - Fires when a dropdown is clicked
```
Fixed in dd20d1bff186b7270ea2ebd6995f3376ef3d8caf

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+    this.id ||= getRandomId('rh-secondary-nav-dropdown');
+
+    const [link] = this.#slots.getSlotted<HTMLElement>('link');
+    link.setAttribute('role', 'button');
```
You are not mistaken.  Would a guard clause with a logger return suffice here?

```js
const [link] = this.#slots.getSlotted<HTMLElement>('link');
const [menu] = this.#slots.getSlotted<HTMLElement>('menu');
if (link === undefined) {
  this.#logger.warn('[rh-secondary-nav-dropdown][slot="link"] expects a slotted <a> tag');
  return;
}
if (menu === undefined) {
  this.#logger.warn('[rh-secondary-nav-dropdown][slot="menu"] expects a slotted <ul> or <ol> tag');
  return;
}
 ```

zeroedin replied:
Fixed in d8a379871a80d1538340ff2c0d890069a481e92a

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+  private _clickHandler(event: MouseEvent) {
+    event.preventDefault();
+    const expanded = !this.expanded;
+    // trigger change event which evokes the mutation on this.expanded
```
You aren't kidding 🙃   Redundant logic, was definitely written while tired or distracted.

Fixed in b0fc9e3ed05996d19da209b0479a06a6b233161d

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+   */
+  #toggleExpand(expand: boolean) {
+    if (expand) {
+      this._container?.classList.add('expanded');
```
Fixed in b0fc9e3ed05996d19da209b0479a06a6b233161d

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-dropdown.ts:
```diff
@@ -0,0 +1,136 @@
+   * @returns {void}
+   */
+  #open(): void {
+    if (this._container?.classList.contains('expanded')) {
```
Fixed in b0fc9e3ed05996d19da209b0479a06a6b233161d

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-menu-section.ts:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @csspart base    - {HTMLSectionElement} container, <section> element
+**/
+@customElement('rh-secondary-nav-menu-section') @pfelement()
```
Fixed in 0507600cb628ab536778cd1ab2bda775da1ff888

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-lightdom.css:
```diff
@@ -0,0 +1,262 @@
+  font-size: var(--rh-secondary-nav-logo-font-size);
+  letter-spacing: 0.18px;
+  line-height: var(--rh-secondary-nav-logo-line-height, 1.25em);
+  margin-left: 1em;
```
I don't have much background in supporting multi-language environments, so this is a good practice to start using for sure!

zeroedin replied:
Fixed in e95ac81f0a678bee7e2e3c088f056742f907078b

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-lightdom.css:
```diff
@@ -0,0 +1,262 @@
+  }
+
+  rh-secondary-nav > [slot="nav"] > li {
+    border-top-width: 0px;
```
Fixec in e95ac81f0a678bee7e2e3c088f056742f907078b

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-menu-section.ts:
```diff
@@ -0,0 +1,86 @@
+ * @slot links      - Adds a ul tag to section, expects `<ul>, <ol>`
+ * @slot cta        - Adds a section level CTA, expects `<pfe-cta>`
+ *
+ * @csspart base    - {HTMLSectionElement} container, <section> element
```
Fixed in 2f88b975241399448e917d6be39fb408e4026d29

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-menu.css:
```diff
@@ -0,0 +1,102 @@
+:host {
+  display: contents;
```
Fixed in d3516a1bf3183a7c3ad3590e3311b1e356701e6d

coreyvickery commented:
## 6/14 feedback

### Styling
* The Dark theme CTA arrow is not changing color on hover ([watch video](https://drive.google.com/file/d/1JwOXNdJ4JReMrdoKW4o3h158CRRFWwUA/view?usp=sharing))
* It might be my weird computer (I swear it was working correctly earlier), but some styles in both themes are not being picked up; worth looking into just to be safe ([watch video](https://drive.google.com/file/d/1adVTE7Mptb56cd0lnC-hKbJ2k6zfvXCi/view?usp=sharing))
<img width="965" alt="Screen Shot 2022-06-14 at 2 31 32 AM" src="https://user-images.githubusercontent.com/45767595/173508861-0253d029-2573-4184-99d1-412cf7bc47fe.png">
<img width="878" alt="Screen Shot 2022-06-14 at 2 38 57 AM" src="https://user-images.githubusercontent.com/45767595/173509999-a896355f-1537-45d9-8ec5-06d277e43b93.png">

### Accessibility
* After I close the menu, the red hover line is still visible. Is this because I'm still technically hovering over the menu I just closed? ([watch video](https://drive.google.com/file/d/19dfhnhUJZG6wuV_zRU_JE20A0b1ygek3/view?usp=sharing))
* Possible to click or tap on the component background to close the menu? If not, no worries ([watch video](https://drive.google.com/file/d/1JLbWJczPTy6cElYlTB69xL9163RT5RxV/view?usp=sharing))
* There is still no hover state for the product name text in both themes (XD has the correct styles)
* Whether using a cursor or keyboard (and in addition to the above request to be able to click or tap on the component background), can we have the `esc` key close the menu as well? (Pressing `enter` opens the menu and pressing `enter` again closes the menu if focus stays where it is, which is good)
* Still seeing the below when hovering on tablet and mobile sizes
<img width="900" alt="Screen Shot 2022-06-14 at 2 10 20 AM" src="https://user-images.githubusercontent.com/45767595/173505615-5262d084-c80f-4877-93f3-b4128d91bb38.png">

### Spacing
* Possible to reduce the amount of spacing between the title (`Use cases`) and bottom edge to 24px?
<img width="865" alt="Screen Shot 2022-06-14 at 2 12 38 AM" src="https://user-images.githubusercontent.com/45767595/173505993-f0e1c06f-d2f2-4776-890c-8732b6654bf2.png">

### Other
* How are we feeling about so many scrollbars? I would like to hear about how others feel about scrolling within an accordion panel as well as a menu. This could also get screwy on mobile when content definitely scrolls a lot, let's keep discussing this
<img width="895" alt="Screen Shot 2022-06-14 at 2 16 09 AM" src="https://user-images.githubusercontent.com/45767595/173506445-7cebca4c-7d43-4ad5-a8c6-8b87ccf3f1c8.png">

Thanks so much, looking really good.

zeroedin commented:
> The Dark theme CTA arrow is not changing color on hover
Fixed in e0088a4d0828ed87d9ae632e93330c9a28d0a4aa

> It might be my weird computer (I swear it was working correctly earlier), but some styles in both themes are not being picked up; worth looking into just to be safe
We discovered this was an issue with not having the latest branch updates.  

> After I close the menu, the red hover line is still visible. Is this because I'm still technically hovering over the menu I just closed?
The menu has focus, this is a focus style being applied, this looks to be correct per the accessibility page in the design spec.  Let me know if that needs to be changed. 

> Possible to click or tap on the component background to close the menu? If not, no worries
Possible, yes, but I wonder if it is desirable as far as accessibility concerns go with reduced dexterity users and the mouse if creating too large of an area that would accidentally close the menu.  I wouldn't love to have @markcaron and @kelsS weigh in on this issue and find out if it adds a benefit or detriment to users.

> Whether using a cursor or keyboard (and in addition to the above request to be able to click or tap on the component background), can we have the esc key close the menu as well? (Pressing enter opens the menu and pressing enter again closes the menu if the focus stays where it is, which is good)
I don't think this is unreasonable but would like @markcaron and @kelsS to weigh in here.  We left out adding additional keyboard navigation as the menu itself should be navigated as a semantic UL of links.  Should we add in `esc` keypress event listener and manage focus by placing it on the "last open" menu item button after the menu is closed as an expected default?

> Still seeing the below when hovering on tablet and mobile sizes
Fixed in a005f23687bb80eb133afa65f346c6cae25484ce

> Possible to reduce the amount of spacing between the title (Use cases) and bottom edge to 24px?
Fixed in 25f2260596bb395c3c043418fd44de52ab949ec8

> How are we feeling about so many scrollbars? I would like to hear about how others feel about scrolling within an accordion panel as well as a menu. This could also get screwy on mobile when content definitely scrolls a lot, let's keep discussing this
It's definitely not my favorite.  Going to take a look closer at how current `rh-navigation` handles the overflow/scrolling.

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-menu-section.ts:
```diff
@@ -0,0 +1,86 @@
+  render() {
+    return html`
+      <section part="base">
+        <slot name="base">
```
Fixed in dae47d21f1dbb493c0ea02097e84efb399fe5e64

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-menu-section.ts:
```diff
@@ -0,0 +1,86 @@
+ * @param tagName {String}
+ * @returns {boolean}
+ */
+// TODO: Rh-footer uses this function as well, should be abstracted into functions. (pfe-core/functions?)
```
I moved the function to `/lib` in 400704e4a3c413a1a0751e330368d1eabf7ee918
I implemented the type predicate in 4fef40481478723f2aaf90fa32b85b5946fe246a

Let me know if the type predicate solution is what you were looking.

bennypowers commented on lib/functions.ts:
```diff
@@ -1,9 +1,7 @@
  */
-export function isHeader(tagName: string): boolean {
-  return !!tagName.match(/^H[1-6]$/i);
+export function isHeadingElement(node: Node | null): node is HTMLHeadingElement {
```
[TIL](https://developer.mozilla.org/en-US/docs/Web/API/HTMLHeadingElement) thanks!

I didn't know this class exists. If you want, you can remove this and just inline the instanceof check, as far as TypeScript is concerned, it's the same thing. You can keep this predicate here if you prefer, too.

zeroedin commented on elements/rh-secondary-nav/rh-secondary-nav-lightdom.css:
```diff
@@ -0,0 +1,262 @@
+rh-secondary-nav {
+  --rh-secondary-nav-min-height: 4.625em;
```
Modified this approach by representing some of these CSS props as "private"  although they are still exposed in lightdom.css for critical CSS needs.  Changes in 66c0fb8444aa8747f4421325959617c674a37201

I've added token values for colors, but I have not approached `<length>` yet due to open discussions on `px` vs `r/em` on the tokens themselves. 

https://github.com/RedHat-UX/red-hat-design-tokens/discussions/12

kelsS commented:
@zeroedin below are the accessibility remediation items for the secondary nav.

Remediation items:

  The `esc` key needs to close the secondary nav menu completely
    -   focus should go back the "menu" button
    -   When secondary nav loses focus the menu should close automatically if it is open
         - The requirements from https://www.w3.org/WAI/ARIA/apg/example-index/disclosure/disclosure-navigation.html(https://www.w3.org/WAI/ARIA/apg/example-index/disclosure/disclosure-navigation.html) for the navigation menu need to be implemented
  axe-devtool errors need to be fixed (to see run the axe devtools extension on the component demo.)
  
I am happy to answer any questions and/or walk through the remediation items with you in the Web Accessibility office hours. 🙂

bennypowers commented on elements/rh-secondary-nav/demo/proxy.html:
```diff
@@ -13,11 +13,7 @@
               <h3 slot="header" id="ansible-platform">
                 <a href="#">Why Red Hat Ansible Automation Platforms</a>
               </h3>
-<<<<<<< HEAD
```
resolve please

bennypowers commented on elements/rh-secondary-nav/rh-secondary-nav.ts:
```diff
@@ -436,6 +416,14 @@ export class RhSecondaryNav extends LitElement {
       this.style.setProperty('--_chevron-transform-expanded', 'rotate(var(--_chevron-down-angle)) translate(var(--_chevron-thickness), calc(-1 * var(--_chevron-thickness)))');
```
Think we could set this on a shadow container?

zeroedin replied:
@bennypowers https://github.com/RedHat-UX/red-hat-design-system/pull/259/commits/b78be73d8eb8f5536bf8b98c1f2d8be054897f7c this change work for you?

bennypowers replied:
yup

bennypowers commented:
Great work, @zeroedin and @kelsS. Very impressive!
